bpf: Add stateless RFC8215 NAT46/64 for standalone lb

borkmann opened the pull request:
see commit msg

borkmann committed:
bpf: Add stateless RFC8215 NAT46/64 for standalone lb

Implement stateless RFC8215 NAT46/64 translation, so this can be used as a
gateway at the edge of the K8s cluster e.g. along with DNS64.

Example:

  # ./daemon/cilium-agent --enable-ipv4=true --enable-ipv6=true \
    --datapath-mode=lb-only --bpf-lb-algorithm=maglev --bpf-lb-maglev-table-size=2039 \
    --bpf-lb-acceleration=disabled --devices=enp5s0 --disable-envoy-version-check=true \
    --enable-stateless-nat46x64=true

-> Transparent NAT46:

  On cluster node (setup of next hop & curl):

  # ip r add 1.1.1.2 via 192.168.2.11 dev enp5s0
  # curl 1.1.1.2:8080

  On gateway node, NAT46:

  # tcpdump -i enp5s0 port 8080 -n
  [...]
  09:59:54.533098 IP 192.168.2.12.43336 > 1.1.1.2.8080: Flags [S], seq 2273864075, win 42340, options [mss 1460,sackOK,TS val 1302308726 ecr 0,nop,wscale 9], length 0
  09:59:54.533122 IP6 64:ff9b::c0a8:20c.43336 > 64:ff9b::101:102.8080: Flags [S], seq 2273864075, win 42340, options [mss 1460,sackOK,TS val 1302308726 ecr 0,nop,wscale 9], length 0
  09:59:54.533262 IP6 64:ff9b::101:102.8080 > 64:ff9b::c0a8:20c.43336: Flags [R.], seq 0, ack 2273864076, win 0, length 0
  09:59:54.533269 IP 1.1.1.2.8080 > 192.168.2.12.43336: Flags [R.], seq 0, ack 2273864076, win 0, length 0
  [...]

  The K8s cluster can expose [64:ff9b::101:102]:8080 or other addresses with a
  64:ff9b::/96 prefix as a LoadBalancer service. The stateless NAT46/64 gateway
  does not need to have awareness of specific VIP:ports. Routing needs to ensure
  that traffic towards the public IPv4 addresses point to NAT46/64 gateway.

-> Transparent NAT64:

  On cluster node (setup of next hop & curl):

  K8s cluster aka CNI would have to make sure that phys ifaces get a
  64:ff9b::/96 address with a public IPv4 pool. Agent needs to install the
  /96 route to the dual-stack standalone LB as next hop with src address
  selection of the publicly routable IPv4. Manual demo example:

  # ip -6 addr add 64:ff9b::101:102/128 dev enp5s0
  # ip -6 r add 64:ff9b::/96 via 2a02:168:f656:0:11e1:2dff:fe01:c164 src 64:ff9b::101:102 dev enp5s0
  # curl "[64:ff9b::c0a8:20c]:8080"

  On gateway node, NAT64:

  # tcpdump -i enp5s0 port 8080 -n
  [...]
  10:04:50.071047 IP6 64:ff9b::101:102.57980 > 64:ff9b::c0a8:20c.8080: Flags [S], seq 3795526123, win 43200, options [mss 1440,sackOK,TS val 2284871983 ecr 0,nop,wscale 9], length 0
  10:04:50.071072 IP 1.1.1.2.57980 > 192.168.2.12.8080: Flags [S], seq 3795526123, win 43200, options [mss 1440,sackOK,TS val 2284871983 ecr 0,nop,wscale 9], length 0
  10:04:50.071262 IP 192.168.2.12.8080 > 1.1.1.2.57980: Flags [R.], seq 0, ack 3795526124, win 0, length 0
  10:04:50.071284 IP6 64:ff9b::c0a8:20c.8080 > 64:ff9b::101:102.57980: Flags [R.], seq 0, ack 3795526124, win 0, length 0
  [...]

Tested with XDP & tc BPF.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
Link: https://www.youtube.com/watch?v=z636zokC_WI

tommyp1ckles reviewed (approved):
agent changes lgtm

**edit**: not sure why it approved sig-datapath on my behalf, I should be excluded from reviews for that one. You'll want to assign someone else for those changes.

borkmann committed:
cilium, helm: Add stateless RFC8215 NAT46/64 knob

Add knob for it, so it can be used from Helm. Also add small runtime test
to nat46x64 CI, more tbd.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

nebril commented on test/nat46x64/test.sh:
```diff
@@ -305,6 +305,34 @@ CILIUM_POD_NAME=$(kubectl -n kube-system get pod -l k8s-app=cilium -o=jsonpath='
+    --set statelessNat46x64.enabled=true
+kubectl -n kube-system delete pod -l k8s-app=cilium
+kubectl -n kube-system rollout status ds/cilium --timeout=5m
+
```
Should there be any test scenarios here aside from changing Cilium config?

borkmann replied:
The translation logic is already checked in earlier unit test, but I'm following up on a stateless test case after KubeCon (no cycles till then).

nebril reviewed (commented):
Agent changes look good to me, I am not sure how the added `helm upgrade` bits test the feature aside from making sure that enabling the feature doesn't crash Cilium agent. Is it enough testing?

aspsk commented on bpf/lib/nat_46x64.h:
```diff
@@ -344,4 +370,29 @@ static __always_inline int ipv6_to_ipv4(struct __ctx_buff *ctx,
 
+static __always_inline int
+nat46_rfc8215(struct __ctx_buff *ctx __maybe_unused,
+	      const struct iphdr *ip4 __maybe_unused,
```
Why do we mark such arguments as unused (not only here, in many places)?

borkmann replied:
Otherwise there's a compiler warning which afaik throws an error Wunused or such. Maybe we should just remove these in general given they've been more churn than 'improvement'.

aspsk replied:
But it _is_ used :-) Why a compiler would blame?

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -1729,7 +1857,12 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
 		if (svc)
 			return DROP_IS_CLUSTER_IP;
-
+#ifdef ENABLE_NAT_46X64_STATELESS
+		if (ip4->daddr != IPV4_DIRECT_ROUTING) {
+			ep_tail_call(ctx, CILIUM_CALL_IPV46_RFC8215);
+			return DROP_MISSED_TAIL_CALL;
+		}
+#endif
```
@borkmann I didn't look too deep. But don't we want this to be in scope of the `skip_service_lookup` label below, so that ICMP traffic also gets 46-translated?

borkmann replied:
@julianwiedmann good point, I was focused only on TCP/UDP, but our nat46x64 code also supports icmp. I'm currently working on the code anyway, so will fix it up along the way. Thanks for spotting!
